### PR TITLE
`@defichain/jellyfish-testing` new centralized testing framework for jellyfish ecosystem

### LINKED ISSUE #431: `@defichain/testing` QOL utilities and setups
On the point of not using existing `client.xx.xx` where they are already `typed`, we don't need to be so strict about it. The best-case scenario would total isolation of concerns in each test case (file in this case) but that would just be overkill. If there is an existing typed method from the client. You can just use it. But for some obvious scenarios where we need isolation, e.g. it's not yet typed but we need assert it.

TLDR: it's okay to use `client.account.getAccount` if it's already merged instead of `container.call('getaccount', [])`.

> We need to look into this for greater testing QOL, but right now we don't have resources for it.

/area testing
/kind refactor

_Originally posted by @fuxingloh in https://github.com/DeFiCh/jellyfish/issues/399#issuecomment-870596525_

### PR DESCRIPTION
**Pull request by fuxingloh**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

This package introduces a centralized testing framework for the jellyfish replacing the existing `@defichain/testing`.
All methods within `@defichain/testing` have thus been deprecated.

Testing is an essential part of any serious quality software developer work. This package provides many abstractions for
various commonly used setup patterns for the DeFi blockchain. This keeps your testing setup DRY for repeated tests.

`@defichain/jellyfish-testing` relies on other jellyfish dependencies, they are cyclic dependant. However, within this
mono-repo, we don't need to declare `"devDependencies"` as they are mapped in `tsconfig.base.json`. This configuration
allows any packages within `jellyfish-*` to rely on `jellyfish-testing` without causing cyclic dependencies errors.

This new implementation warps a container on init via `Testing.create(container)` extending on top of rpc and
container. The expectations are other more complex fixtures will be added into `testing.fixture`. 

```ts
const container = new MasterNodeRegTestContainer()
const testing = Testing.create(container)

testing.rpc         // to access JSON RPC
testing.fixture     // for complex fixture setup e.g. poolpair
testing.rawtx       // for rawtx setup with reasonable defaults 
testing.token       // for token setup with reasonable defaults 
testing.poolpair    // for poolpair setup with reasonable defaults 
```

On the point of not using existing JSON RPC within tests, this is optional moving forward, please use them to keep it
DRY.

#### Which issue(s) does this PR fixes?:

Fixes #431
Fixes #566 with `testing.rawtx.getDfTx(txid)`

#### Additional comments

In this PR, I only replaced certain setups for brevity, we will deprecate and slowly move towards using `jellyfish-testing` reclusively in days to come.

### CODE REVIEW

File: packages/jellyfish-testing/src/fixture.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * TestingFixture setup complex fixtures for jellyfish testing.
+ */
+export class TestingFixture {
+  constructor (
+    private readonly testing: Testing
+  ) {
+  }
+
+  async createPoolPair (options: TestingPoolPairCreateRequest): Promise<poolpair.PoolPairInfo> {
+    if (options.a.symbol !== 'DFI') {
```
Comment (jingyi2811) [inline packages/jellyfish-testing/src/fixture.ts:14]: Is that better if you have `await this.testing.generate(1)` for every calls?

File: packages/jellyfish-testing/src/rawtx.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  async sign (options: TestingRawTxSign): Promise<string> {
+    const { hex, privKeys } = options
+
+    const mapped = await asPrivKey(privKeys)
+    const signed = await this.rpc.rawtx.signRawTransactionWithKey(hex, mapped)
+    return signed.hex
+  }
+
+  /**
+   * @param {string} txid to of a DfTx
```
Comment (jingyi2811) [inline packages/jellyfish-testing/src/rawtx.ts:45]: Maybe you can remove the comment section like other functions in this file.
If not, the return type should be `Promise<TestingRawTxDfTx>`

File: packages/jellyfish-testing/src/fixture.ts (hunk 1 above)
Comment (canonbrother) [reply packages/jellyfish-testing/src/fixture.ts:14]: hmm.. seems its intention, is tit better manageable for multi-nodes??

Comment (fuxingloh) [reply packages/jellyfish-testing/src/fixture.ts:14]: Yeah, it is intentional as block generation is quite expensive time-wise. (You can't generate too far into the future)

By design `fixture.ts` should only generate blocks they need, so we don't waste expensive testing cycle. As this is a shared fixture for the entirety of the jellyfish testing ecosystem.

**Review by canonbrother — APPROVED** [review]

**Review by jingyi2811 — APPROVED** [review]

**Review by surangap — APPROVED** [review]
nice :clap:

**Pull request merged**